packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -35,6 +35,70 @@ const int _kPubExitCodeUnavailable = 69;
 
 typedef MessageFilter = String? Function(String message);
 
+/// globalCachePath is the directory in which the content of the localCachePath will be moved in
+bool joinCaches({
+  required FileSystem fileSystem,
+  required String globalCachePath,
+  required String localCachePath,
+  bool start = true,
+}) {
+  final Directory globalDirectory = fileSystem.directory(globalCachePath);
+  final Directory localDirectory = fileSystem.directory(localCachePath);
+
+  for (final FileSystemEntity entity in localDirectory.listSync()) {
+    final String newPath = fileSystem.path.join(globalDirectory.path, entity.basename);
+    if (entity is File) {
+      try {
+        entity.copySync(newPath);
+      }
+      on FileSystemException {
+        if (!start) { // check if it's not the first iteration; we only want to delete newly created directories
+          globalDirectory.deleteSync(recursive: true);
+        }
+        return false;
+      }
+    } else if (entity is Directory) {
+      if (!globalDirectory.childDirectory(entity.basename).existsSync()) {
+        final Directory newDirectory = globalDirectory.childDirectory(entity.basename);
+        try {
+          newDirectory.createSync();
+        }
+        on FileSystemException {
+          if (!start) {
+            globalDirectory.deleteSync(recursive: true);
+          }
+          return false;
+        }
+        return joinCaches(
```

Comment:
I think this is wrong, I think by early returning we will never visit all of the entities in `localDirectory.listSync()` on line 48